src/Promotions/InPluginUpsells/SaleBanners.php

```diff
@@ -161,4 +207,78 @@ public static function isShowing(): bool
     {
         return isset($_GET['post_type']) && $_GET['post_type'] === 'give_forms';
     }
+
+    /**
+     * @unreleased retrieve licensed plugin slugs.
+     */
+    public static function getLicensedPluginSlugs(): array
+    {
+        $pluginSlugs = [];
+        $licenses = get_option("give_licenses", []);
+
+        foreach ($licenses as $license) {
+            if (isset($license['is_all_access_pass']) && $license['is_all_access_pass'] && !empty($license['download'])) {
+                $pluginSlugs = ['is_all_access_pass'];
+            } else {
+                $pluginSlugs[] = $license['plugin_slug'];
+            }
+        }
+
+        return $pluginSlugs;
+    }
+
+    /**
+     * @unreleased determines user pricing plan from licensed plugin slugs.
+     */
+    public static function getUserPricingPlan(): string
+    {
+        $plan = 'Free';
+
+        $basic = [
+            'pdf' => 'give-pdf-receipts',
+        ];
+
+        $plus = [
+            'pdf_receipts' => 'give-pdf-receipts',
+            'recurring_donations' => 'give-recurring',
+            'fee_recovery' => 'give-fee-recovery',
+            'form_field_manager' => 'give-form-field-manager',
+            'tributes' => 'give-tributes',
+            'annual_receipts' => 'give-annual-receipts',
+            'peer_to_peer' => 'give-peer-to-peer',
+        ];
+
+        $pro = ['is_all_access_pass'];
+
+        $licensedPluginSlugs = self::getLicensedPluginSlugs();
+
+        sort($licensedPluginSlugs);
+        sort($basic);
+        sort($plus);
+
+        if ($licensedPluginSlugs === $basic) {
+            $plan = 'Basic';
+        } elseif ($licensedPluginSlugs === $plus) {
+            $plan = 'Plus';
+        } elseif ($licensedPluginSlugs === $pro) {
+            $plan = 'Pro';
+        }
+
+        return $plan;
+    }
```

Comment:
Hey @JoshuaHungDinh! What if a user has the Plus license but has also purchased an individual license for another Pro plugin? In such a case, I believe that when comparing the licenses, if they are **equal** to one of the plans, we will consider the user as a Free user.

If my assumption is correct, we would need to modify the comparison approach. Instead of simply checking for equality, we would need to verify if the user has obtained all the licenses associated with that plan. This can be achieved by using the array diff method:

![CleanShot 2023-11-01 at 17 00 16](https://github.com/impress-org/givewp/assets/3921017/ea2fc47b-d5ff-4369-b036-c935b9a4c9b9)

Btw, sorting is not necessary when using `array_diff_assoc`.

Reply:
Resolved: 0ab19add5d816fedf4b5415ae9cc2cb2dcb63e41